File: app/src/commands/market.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from "@tauri-apps/api/core";
import type {
  BuyResourcesRequest,
  GetMarketFeeRequest,
  GetMarketFeeResponse,
  GetMarketRequest,
  GetMarketResponse,
  Resources,
  Ruler,
  SellResourcesRequest,
  SendResourcesRequest,
} from "@tsukilabs/nil-bindings";

export async function buyResources(resources: Resources) {
  const req: BuyResourcesRequest = {
    world: NIL.world.getIdStrict(),
    resources,
  };

  await invoke("buy_resources", { req });
}

export async function getMarket() {
  const req: GetMarketRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetMarketResponse>("get_market", { req });
}

export async function getMarketFee() {
  const req: GetMarketFeeRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetMarketFeeResponse>("get_market_fee", { req });
}

export async function sellResources(resources: Resources) {
  const req: SellResourcesRequest = {
    world: NIL.world.getIdStrict(),
    resources,
  };

  await invoke("sell_resources", { req });
}

export async function sendResources(recipient: Ruler, resources: Resources) {
  const req: SendResourcesRequest = {
    world: NIL.world.getIdStrict(),
    recipient,
    resources,
  };

  await invoke("send_resources", { req });
}

File: app/src-tauri/src/command/market.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::manager::ManagerExt;
use nil_payload::request::market::*;
use nil_payload::response::market::*;
use tauri::AppHandle;

#[tauri::command]
pub async fn buy_resources(app: AppHandle, req: BuyResourcesRequest) -> Result<()> {
  app
    .client(async |cl| cl.buy_resources(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_market(app: AppHandle, req: GetMarketRequest) -> Result<GetMarketResponse> {
  app
    .client(async |cl| cl.get_market(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_market_fee(
  app: AppHandle,
  req: GetMarketFeeRequest,
) -> Result<GetMarketFeeResponse> {
  app
    .client(async |cl| cl.get_market_fee(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn sell_resources(app: AppHandle, req: SellResourcesRequest) -> Result<()> {
  app
    .client(async |cl| cl.sell_resources(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn send_resources(app: AppHandle, req: SendResourcesRequest) -> Result<()> {
  app
    .client(async |cl| cl.send_resources(req).await)
    .await
    .map_err(Into::into)
}
